### PR TITLE
 expose input variables that only shape needed in each subgraph that compiled by CINN

### PR DESCRIPTION
**Pull request by CtfGo**
### PR types

Performance optimization 
### PR changes

OPs

### Describe

collecting input variables that only shape needed of  each subgraph that compiled by CINN in `build_cinn_pass`, and expose them to memory optimization of framework passes by declaring`DECLARE_INPLACE_OP_INFERER` in `cinn_launch` op.

### CODE REVIEW

File: paddle/fluid/framework/paddle2cinn/build_cinn_pass.cc [hunk 1]
```diff
@@ -214,6 +215,37 @@ void AddOutputVar(const GraphNodeSet& output_vars, const GraphNodeSet& cluster,
...
+    const GraphNodeSet& cluster, const GraphNodeSet& cluster_inputs) {
+  // 1. Find ops with NoNeedBufferVarsInferer and collect their argument names
+  std::unordered_set<std::string> no_need_buffer_args;
+  for (const auto* op_node : cluster) {
+    auto& inferer =
+        OpInfoMap::Instance().Get(op_node->Name()).NoNeedBufferVarsInferer();
+    if (!inferer) {
+      continue;
+    }
+    auto* op_desc = op_node->Op();
```
Comment (thisjiang) [inline paddle/fluid/framework/paddle2cinn/build_cinn_pass.cc:228]: 这儿建议加个`Paddle_enfore`判断下op_desc是否为空

File: paddle/fluid/framework/paddle2cinn/build_cinn_pass.cc [hunk 2]
```diff
@@ -214,6 +215,37 @@ void AddOutputVar(const GraphNodeSet& output_vars, const GraphNodeSet& cluster,
...
+    auto no_need_buffer_params =
+        inferer(op_desc->Inputs(), op_desc->Inputs(), op_desc->GetAttrMap());
+    for (const auto& param : no_need_buffer_params) {
+      const auto& args = op_desc->Input(param);
+      no_need_buffer_args.insert(args.begin(), args.end());
```
Comment (thisjiang) [inline paddle/fluid/framework/paddle2cinn/build_cinn_pass.cc]: 感觉这个循环与下面的循环可以合并啊？在这儿加个判断
```
for(auto* arg : op_desc->Input(param)) {
  if (cluster_inputs.count(arg)) {
    ...
  }
}
```

File: paddle/fluid/framework/paddle2cinn/build_cinn_pass.cc (hunk 1 above)
Comment (CtfGo) [reply paddle/fluid/framework/paddle2cinn/build_cinn_pass.cc:228]: fix it

File: paddle/fluid/framework/paddle2cinn/build_cinn_pass.cc (hunk 2 above)
Comment (CtfGo) [reply paddle/fluid/framework/paddle2cinn/build_cinn_pass.cc]: 这里 arg 只是 string，cluster_inputs 是 Node*集合，不能直接查找

**Review by thisjiang — APPROVED** [review]
Great!

**Pull request merged**